include/ratelimit_manager.h

```diff
@@ -0,0 +1,174 @@
+#pragma once
+
+#include <string>
+#include <vector>
+#include <tuple>
+#include <unordered_map>
+#include <mutex>
+#include <shared_mutex>
+#include "lru/lru.hpp"
+#include "option.h"
+
+
+
+
+// Action enum for rate limit rules
+enum class RateLimitAction {
+    ALLOW,
+    BLOCK,
+    THROTTLE
+};
+
+enum class RateLimitedResourceType {
+    IP,
+    API_KEY
+};
+
+struct rate_limit_throttle_t {
+    int64_t minute_rate_limit;
+    int64_t hour_rate_limit;
+
+};
+
+struct rate_limit_rule_t {
+    uint64_t id;
+    RateLimitAction action;
+    RateLimitedResourceType resource_type;
+    std::vector<std::string> values;
+    rate_limit_throttle_t throttle;
+    int64_t auto_ban_threshold_num = -1;
+    int64_t auto_ban_num_days = -1;
+};
+
+// Entry struct for rate limit rule pointer hash map as key
+struct rate_limit_rule_entry_t {
```

Comment:
Let's call this `rate_limit_resource_t` as the `rule_entry` part makes it seem like it's an entry in the rule store, which it is not.